### PR TITLE
qwen2.5vl: fix bugs when using flash2+bf16 or num_return_sequences>1

### PR DESCRIPTION
**Pull request by gewenbin0992**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
Hey @gewenbin0992 !

Just a small note, the precision thing with FA2 was fixed by https://github.com/huggingface/transformers/pull/36065. For the rest of changes, a small reproducer +description of problem will be needed before reviewing :)

**Comment by gewenbin0992** [comment]
```py
import json
import torch
from transformers import Qwen2_5_VLForConditionalGeneration, AutoProcessor
from qwen_vl_utils import process_vision_info


# default: Load the model on the available device(s)
model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen2.5-VL-7B-Instruct", torch_dtype="auto", device_map="auto"
)

# default processor
processor = AutoProcessor.from_pretrained("Qwen/Qwen2.5-VL-7B-Instruct")

messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image",
                "image": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg",
            },
            {"type": "text", "text": "Describe this image."},
        ],
    }
]

# Preparation for inference
text = processor.apply_chat_template(
    messages, tokenize=False, add_generation_prompt=True
)
image_inputs, video_inputs = process_vision_info(messages)

inputs = processor(
    text=[text],
    images=image_inputs,
    videos=video_inputs,
    padding=True,
    return_tensors="pt",
)
inputs = inputs.to(model.device)

num_return_sequences = 3

# Inference: Generation of the output
generated_ids = model.generate(**inputs, max_new_tokens=128, num_return_sequences=num_return_sequences)

input_ids_pxpended = torch.cat([inputs.input_ids] * num_return_sequences)
generated_ids_trimmed = [
    out_ids[len(in_ids) :] for in_ids, out_ids in zip(input_ids_pxpended, generated_ids)
]
output_text = processor.batch_decode(
    generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
)
print(json.dumps(output_text, indent=4))
```
# before these changes, the output was bad
[*because torch.repeat_interleave disrupted the order of tensors without a batch size dimension*](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L744)
```json
[
    "The image is a blank white background with no visible content or text. There are no discernible objects, colors, or patterns present in the image.",
    "The image appears to be a collage or a series of fragmented, pixelated sections that seem to depict a person's hand reaching out towards the sky. The background is a bright, overexposed sky with some indistinct shapes and colors that could be interpreted as clouds or abstract forms. The overall effect is glitchy and distorted, giving the impression of digital interference or a broken image. The hand is positioned in the center, with fingers extended upwards, suggesting a gesture of reaching or aspiration.",
    "The image appears to be a collage or a series of fragmented, pixelated sections that seem to depict a person wearing a plaid shirt and holding a camera. The sections are arranged in a grid-like pattern, with each section showing different parts of the person's body and attire. The overall effect is abstract and glitchy, giving the impression of digital distortion or interference."
]
```
# after these changes, the output is good
```json
[
    "The image depicts a person sitting on the sandy beach, interacting with a light-colored dog. The person is wearing a plaid shirt and dark pants, and they appear to be giving the dog a treat or a small object. The dog is wearing a harness with colorful paw prints and is sitting upright, reaching up with its front paws towards the person's hand. The background shows the ocean with gentle waves and a bright, golden sunset, creating a warm and serene atmosphere. The scene suggests a moment of bonding between the person and their pet in a peaceful outdoor setting.",
    "The image depicts a person sitting on the sandy beach, interacting with a light-colored dog. The person is wearing a plaid shirt and dark pants, and they appear to be giving the dog a treat or a small object. The dog is sitting upright and extending its paw towards the person's hand, seemingly in response to a command or gesture. The background shows the ocean with gentle waves, and the sun is setting, casting a warm, golden glow over the scene. The overall atmosphere is serene and joyful, suggesting a moment of bonding between the person and their pet.",
    "The image depicts a person sitting on the sandy beach, interacting with a light-colored dog. The person is wearing a plaid shirt and dark pants, and they appear to be giving the dog a treat or a small object. The dog is sitting upright and extending its paw towards the person's hand, seemingly in response to a command or gesture. The background shows the ocean with gentle waves, and the sun is setting, casting a warm, golden glow over the scene. The overall atmosphere is serene and joyful, suggesting a moment of bonding between the person and their pet."
]
```
@zucchini-nlp please review

**Comment by gewenbin0992** [comment]
if num_return_sequences > 1, this bug will be triggered.

**Review by zucchini-nlp — COMMENTED** [review]
Hmm, this seems quite tricky if qwen2-vl has no batch dimension. We need to make sure that it actually works with batched inputs. From first glance seems like there returned outputs will be in different order compared to other model, since we do `cat` and not `interleave`

Also, in beam search scenarios we usually sample beams from within one batch assuming the inputs are interleaved. Again not very sure if the changes will make sense for beam search

Can we make sure it works in batch scenarios as expected and add a small test?

**Comment by gewenbin0992** [comment]
@zucchini-nlp thanks for you review, i got it, we should not do cat and not interleave. 

the key point is we need to do the right repeat_interleave by samples

1. if there is a batch dimension, torch.repeat_interleave works fine
2. if there is **no batch dimension**, we need to split the tensor into samples, then do repeat_interleave by samples

updated, please review again : )

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 1]
```diff
@@ -1931,5 +1931,118 @@ def prepare_inputs_for_generation(
...
+                    dict_to_expand[key] = _repeat_interleave_samples(
+                        dict_to_expand[key], lengths=lengths, repeat_times=expand_size
+                    )
+                elif key == "video_grid_thw":
+                    lengths = list(video_nums)
+                    dict_to_expand[key] = _repeat_interleave_samples(
+                        dict_to_expand[key], lengths=lengths, repeat_times=expand_size
+                    )
+                elif key == "second_per_grid_ts":
+                    assert isinstance(dict_to_expand[key], list)
```
Comment (zucchini-nlp) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:2022]: Instead of assert, I'd prefer an explicit way to catch and raise informative errors

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 2]
```diff
@@ -1931,5 +1931,118 @@ def prepare_inputs_for_generation(
...
         return model_inputs
 
+    def _get_image_nums_and_video_nums(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        image_grid_thw: Optional[torch.LongTensor] = None,
+        video_grid_thw: Optional[torch.LongTensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Args:
```
Comment (zucchini-nlp) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1941]: Can we add a small explanation, what does this meth do?

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 3]
```diff
@@ -1931,5 +1931,118 @@ def prepare_inputs_for_generation(
...
+        vision_start_token_id = self.config.vision_start_token_id
+
+        image_nums, video_nums = None, None
+        if input_ids is not None and (image_grid_thw is not None or video_grid_thw is not None):
+            vision_start_mask = input_ids == vision_start_token_id
+            vision_first_mask = torch.roll(vision_start_mask, shifts=1, dims=1)
+            image_mask = input_ids == image_token_id
+            video_mask = input_ids == video_token_id
+            image_nums = torch.sum(vision_first_mask & image_mask, dim=1)
+            video_nums = torch.sum(vision_first_mask & video_mask, dim=1)
```
Comment (zucchini-nlp) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1965]: this will fail when either image or video is `None`, as of of the `mask` will be also `None`

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for iterating! Looks quite complicated. but I believe there's not much we can do with the way `visuals` are batched in Qwen2-VL.

I left a few comments and we also need a slow test, to make sure generated text makes sense

**Review by ArthurZucker — COMMENTED** [review]
~This is also missing a test, will allow us to better understand what this enables as well~ oups talked too soon!

**Review by ArthurZucker — COMMENTED** [review]
Good feature! This might also help enabling beam search no?

**Comment by gewenbin0992** [comment]
@ArthurZucker Right, we fixed the bug in `_expand_inputs_for_generation`, which is also used for beam searching. And, sorry for the complicated code, we are fixing the failed tests.

**Comment by ArthurZucker** [comment]
It's actually quite good! I would just remove the extra nesting and rather use external functions, but that kind of code is much needed for continious batching so completely understand! It shall leave in `generation/utils` once we standardize well

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -1666,7 +1666,9 @@ def test_generate_from_inputs_embeds(self, _, num_beams):
 
             # There are a few exception patterns in this test:
             # 1 - Some models can't generate without `input_ids`, when `inputs_embeds` are passed
-            requires_inputs_ids = any(model_name in model_class.__name__.lower() for model_name in ["idefics"])
+            requires_inputs_ids = any(
+                model_name in model_class.__name__.lower() for model_name in ["idefics", "qwen2vl", "qwen2_5_vl"]
+            )
```
Comment (minostauros) [inline tests/generation/test_utils.py:1671]: Is this necessary?
Because I could generate well with `inputs_embeds` and generating with `inputs_embeds` is very useful to keep extracted visual information beforhand.
That's why I tried to remove this in https://github.com/huggingface/transformers/pull/35466

Comment (gewenbin0992) [reply tests/generation/test_utils.py]: deleted.

**Comment by gewenbin0992** [comment]
@ArthurZucker @minostauros @zucchini-nlp 
All tests about qwen have passed, please check these failed tests
Can we merge?

**Review by zucchini-nlp — APPROVED** [review]
Thanks a lot for fixing this!

**Review by ArthurZucker — APPROVED** [review]
Kudos and thanks a lot for your maintenance effort!!! 🤗
